File: backend/src/main/java/com/platform/saas/service/AutomationService.java
package com.platform.saas.service;

import com.platform.saas.model.AutomationRule;
import com.platform.saas.model.EventLog;
import com.platform.saas.repository.AutomationRuleRepository;
import com.platform.saas.repository.EventLogRepository;
import com.platform.saas.security.TenantContext;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

/**
 * Service for managing automation rules.
 * Handles CRUD operations and provides access to event logs.
 */
@Service
@RequiredArgsConstructor
@Slf4j
@Transactional
public class AutomationService {

    private final AutomationRuleRepository automationRuleRepository;
    private final EventLogRepository eventLogRepository;

    /**
     * Create a new automation rule.
     * @param rule The automation rule to create
     * @return The created automation rule
     */
    public AutomationRule createRule(AutomationRule rule) {
        UUID tenantId = TenantContext.getTenantId();

        if (tenantId == null) {
            throw new IllegalStateException("Tenant context not set");
        }

        log.info("Creating automation rule for tenant: {}", tenantId);

        // Ensure rule belongs to current tenant
        rule.setTenantId(tenantId);

        // Initialize defaults
        if (rule.getIsActive() == null) {
            rule.setIsActive(true);
        }
        if (rule.getExecutionCount() == null) {
            rule.setExecutionCount(0L);
        }

        AutomationRule savedRule = automationRuleRepository.save(rule);
        log.info("Automation rule created: id={}, name={}, event={}, action={}, tenant={}",
                savedRule.getId(), savedRule.getName(), savedRule.getEventType(),
                savedRule.getActionType(), tenantId);

        return savedRule;
    }

    /**
     * Get an automation rule by ID with tenant isolation.
     * @param ruleId The rule ID
     * @return The automation rule
     */
    @Transactional(readOnly = true)
    public AutomationRule getRule(UUID ruleId) {
        UUID tenantId = TenantContext.getTenantId();

        return automationRuleRepository.findById(ruleId)
                .filter(rule -> rule.getTenantId().equals(tenantId))
                .orElseThrow(() -> new RuntimeException("Automation rule not found: " + ruleId));
    }

    /**
     * Get all automation rules for the current tenant.
     * @return List of automation rules
     */
    @Transactional(readOnly = true)
    public List<AutomationRule> getAllRules() {
        UUID tenantId = TenantContext.getTenantId();
        return automationRuleRepository.findByTenantId(tenantId);
    }

    /**
     * Get active automation rules for the current tenant.
     * @return List of active automation rules
     */
    @Transactional(readOnly = true)
    public List<AutomationRule> getActiveRules() {
        UUID tenantId = TenantContext.getTenantId();
        return automationRuleRepository.findByTenantIdAndIsActive(tenantId, true);
    }

    /**
     * Get automation rules by event type.
     * @param eventType The event type (e.g., "task.status.changed")
     * @return List of automation rules for the event type
     */
    @Transactional(readOnly = true)
    public List<AutomationRule> getRulesByEventType(String eventType) {
        UUID tenantId = TenantContext.getTenantId();
        return automationRuleRepository.findByTenantIdAndEventTypeAndIsActive(tenantId, eventType, true);
    }

    /**
     * Get top executed automation rules for the current tenant.
     * @param limit Maximum number of rules to return
     * @return List of top executed rules
     */
    @Transactional(readOnly = true)
    public List<AutomationRule> getTopExecutedRules(int limit) {
        UUID tenantId = TenantContext.getTenantId();
        return automationRuleRepository.findTopExecutedRules(tenantId, limit);
    }

    /**
     * Update an existing automation rule.
     * @param ruleId The rule ID
     * @param updatedRule The updated rule data
     * @return The updated automation rule
     */
    public AutomationRule updateRule(UUID ruleId, AutomationRule updatedRule) {
        UUID tenantId = TenantContext.getTenantId();

        AutomationRule existing = automationRuleRepository.findById(ruleId)
                .filter(rule -> rule.getTenantId().equals(tenantId))
                .orElseThrow(() -> new RuntimeException("Automation rule not found: " + ruleId));

        // Update fields
        if (updatedRule.getName() != null) {
            existing.setName(updatedRule.getName());
        }
        if (updatedRule.getEventType() != null) {
            existing.setEventType(updatedRule.getEventType());
        }
        if (updatedRule.getActionType() != null) {
            existing.setActionType(updatedRule.getActionType());
        }
        if (updatedRule.getConditions() != null) {
            existing.setConditions(updatedRule.getConditions());
        }
        if (updatedRule.getActionConfig() != null) {
            existing.setActionConfig(updatedRule.getActionConfig());
        }
        if (updatedRule.getIsActive() != null) {
            existing.setIsActive(updatedRule.getIsActive());
        }

        AutomationRule saved = automationRuleRepository.save(existing);
        log.info("Automation rule updated: id={}, tenant={}", ruleId, tenantId);

        return saved;
    }

    /**
     * Delete an automation rule.
     * @param ruleId The rule ID
     */
    public void deleteRule(UUID ruleId) {
        UUID tenantId = TenantContext.getTenantId();

        AutomationRule rule = automationRuleRepository.findById(ruleId)
                .filter(r -> r.getTenantId().equals(tenantId))
                .orElseThrow(() -> new RuntimeException("Automation rule not found: " + ruleId));

        automationRuleRepository.delete(rule);
        log.info("Automation rule deleted: id={}, tenant={}", ruleId, tenantId);
    }

    /**
     * Toggle automation rule active status.
     * @param ruleId The rule ID
     * @param isActive The new active status
     * @return The updated automation rule
     */
    public AutomationRule toggleRuleStatus(UUID ruleId, boolean isActive) {
        UUID tenantId = TenantContext.getTenantId();

        AutomationRule rule = automationRuleRepository.findById(ruleId)
                .filter(r -> r.getTenantId().equals(tenantId))
                .orElseThrow(() -> new RuntimeException("Automation rule not found: " + ruleId));

        rule.setIsActive(isActive);
        AutomationRule saved = automationRuleRepository.save(rule);

        log.info("Automation rule status changed: id={}, active={}, tenant={}",
                ruleId, isActive, tenantId);

        return saved;
    }

    /**
     * Count automation rules for the current tenant.
     * @return The rule count
     */
    @Transactional(readOnly = true)
    public long countRules() {
        UUID tenantId = TenantContext.getTenantId();
        return automationRuleRepository.countByTenantId(tenantId);
    }

    /**
     * Get recent event logs for the current tenant.
     * @param limit Maximum number of logs to return
     * @return List of recent event logs
     */
    @Transactional(readOnly = true)
    public List<EventLog> getRecentLogs(int limit) {
        UUID tenantId = TenantContext.getTenantId();
        return eventLogRepository.findRecentLogs(tenantId, limit);
    }

    /**
     * Get event logs for a specific automation rule.
     * @param ruleId The rule ID
     * @return List of event logs for the rule
     */
    @Transactional(readOnly = true)
    public List<EventLog> getLogsForRule(UUID ruleId) {
        UUID tenantId = TenantContext.getTenantId();

        // Verify rule belongs to tenant
        automationRuleRepository.findById(ruleId)
                .filter(rule -> rule.getTenantId().equals(tenantId))
                .orElseThrow(() -> new RuntimeException("Automation rule not found: " + ruleId));

        return eventLogRepository.findByAutomationRuleId(ruleId);
    }

    /**
     * Get failed event logs for the current tenant.
     * @return List of failed event logs
     */
    @Transactional(readOnly = true)
    public List<EventLog> getFailedLogs() {
        UUID tenantId = TenantContext.getTenantId();
        return eventLogRepository.findByTenantIdAndStatusOrderByCreatedAtDesc(
                tenantId,
                EventLog.ExecutionStatus.FAILED
        );
    }

    /**
     * Get event logs within a date range.
     * @param startDate Start of the date range
     * @param endDate End of the date range
     * @return List of event logs within the date range
     */
    @Transactional(readOnly = true)
    public List<EventLog> getLogsByDateRange(LocalDateTime startDate, LocalDateTime endDate) {
        UUID tenantId = TenantContext.getTenantId();
        return eventLogRepository.findByTenantIdAndDateRange(tenantId, startDate, endDate);
    }

    /**
     * Get count of event logs by status.
     * @param status The execution status
     * @return Count of event logs with the given status
     */
    @Transactional(readOnly = true)
    public long countLogsByStatus(EventLog.ExecutionStatus status) {
        UUID tenantId = TenantContext.getTenantId();
        return eventLogRepository.countByTenantIdAndStatus(tenantId, status);
    }

    /**
     * Get average execution duration for automations.
     * @return Average execution duration in milliseconds
     */
    @Transactional(readOnly = true)
    public Double getAverageExecutionDuration() {
        UUID tenantId = TenantContext.getTenantId();
        Double avg = eventLogRepository.getAverageExecutionDuration(tenantId);
        return avg != null ? avg : 0.0;
    }
}

File: frontend/src/services/automationService.ts
import { apiClient } from './api';

export type ExecutionStatus = 'SUCCESS' | 'FAILED' | 'SKIPPED' | 'NO_RULES_MATCHED';

export interface AutomationRule {
  id?: string;
  tenantId?: string;
  name: string;
  eventType: string;
  actionType: string;
  conditions?: Record<string, any>;
  actionConfig: Record<string, any>;
  isActive?: boolean;
  executionCount?: number;
  lastExecutedAt?: string;
  createdAt?: string;
  updatedAt?: string;
}

export interface EventLog {
  id: string;
  tenantId: string;
  automationRuleId?: string;
  eventType: string;
  actionType?: string;
  status: ExecutionStatus;
  eventPayload?: Record<string, any>;
  actionResult?: Record<string, any>;
  errorMessage?: string;
  errorStackTrace?: string;
  executionDurationMs?: number;
  createdAt: string;
  resourceId?: string;
  resourceType?: string;
}

export interface AutomationStats {
  totalRules: number;
  successfulExecutions: number;
  failedExecutions: number;
  skippedExecutions: number;
  noRulesMatched: number;
  averageExecutionDurationMs: number;
}

export const automationService = {
  // Automation Rules

  /**
   * Create a new automation rule.
   */
  async createRule(rule: AutomationRule): Promise<AutomationRule> {
    const response = await apiClient.post<AutomationRule>('/automations', rule);
    return response.data;
  },

  /**
   * Get all automation rules.
   */
  async getAllRules(activeOnly?: boolean): Promise<AutomationRule[]> {
    const response = await apiClient.get<AutomationRule[]>('/automations', {
      params: { activeOnly },
    });
    return response.data;
  },

  /**
   * Get a specific automation rule by ID.
   */
  async getRule(id: string): Promise<AutomationRule> {
    const response = await apiClient.get<AutomationRule>(`/automations/${id}`);
    return response.data;
  },

  /**
   * Get automation rules by event type.
   */
  async getRulesByEventType(eventType: string): Promise<AutomationRule[]> {
    const response = await apiClient.get<AutomationRule[]>('/automations/by-event-type', {
      params: { eventType },
    });
    return response.data;
  },

  /**
   * Get top executed automation rules.
   */
  async getTopExecutedRules(limit: number = 10): Promise<AutomationRule[]> {
    const response = await apiClient.get<AutomationRule[]>('/automations/top-executed', {
      params: { limit },
    });
    return response.data;
  },

  /**
   * Update an automation rule.
   */
  async updateRule(id: string, rule: Partial<AutomationRule>): Promise<AutomationRule> {
    const response = await apiClient.put<AutomationRule>(`/automations/${id}`, rule);
    return response.data;
  },

  /**
   * Toggle automation rule active status.
   */
  async toggleRuleStatus(id: string, isActive: boolean): Promise<AutomationRule> {
    const response = await apiClient.patch<AutomationRule>(`/automations/${id}/toggle`, {
      isActive,
    });
    return response.data;
  },

  /**
   * Delete an automation rule.
   */
  async deleteRule(id: string): Promise<void> {
    await apiClient.delete(`/automations/${id}`);
  },

  /**
   * Get count of automation rules.
   */
  async getRuleCount(): Promise<number> {
    const response = await apiClient.get<{ count: number }>('/automations/count');
    return response.data.count;
  },

  // Event Logs

  /**
   * Get recent event logs.
   */
  async getRecentLogs(limit: number = 50): Promise<EventLog[]> {
    const response = await apiClient.get<EventLog[]>('/automations/logs', {
      params: { limit },
    });
    return response.data;
  },

  /**
   * Get event logs for a specific automation rule.
   */
  async getLogsForRule(ruleId: string): Promise<EventLog[]> {
    const response = await apiClient.get<EventLog[]>(`/automations/${ruleId}/logs`);
    return response.data;
  },

  /**
   * Get failed event logs.
   */
  async getFailedLogs(): Promise<EventLog[]> {
    const response = await apiClient.get<EventLog[]>('/automations/logs/failed');
    return response.data;
  },

  /**
   * Get event logs within a date range.
   */
  async getLogsByDateRange(startDate: string, endDate: string): Promise<EventLog[]> {
    const response = await apiClient.get<EventLog[]>('/automations/logs/date-range', {
      params: { startDate, endDate },
    });
    return response.data;
  },

  // Statistics

  /**
   * Get automation statistics.
   */
  async getStats(): Promise<AutomationStats> {
    const response = await apiClient.get<AutomationStats>('/automations/stats');
    return response.data;
  },
};
